main/ZactuatorFASTLED.ino

```diff
@@ -0,0 +1,238 @@
+/*
+  OpenMQTTGateway - ESP8266 or Arduino program for home automation
+
+  This gateway enables to use RGBLED strips like WS2812
+
+    Copyright: (c)
+
+    This file is part of OpenMQTTGateway.
+
+    OpenMQTTGateway is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    OpenMQTTGateway is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "User_config.h"
+
+#ifdef ZactuatorFASTLED
+#include <ArduinoJson.h>
+
+#ifdef ESP8266
+#include <FastLED.h>
+#else
+#include <FastLED.h>
+#endif
+enum LEDState {OFF,FIRE,GENERAL};
+LEDState currentLEDState;
+
+CRGB leds[FASTLED_NUM_LEDS];
+CRGB ledColorBlink[FASTLED_NUM_LEDS];
+StaticJsonBuffer<200> jsonBuffer;
+bool blinkLED[FASTLED_NUM_LEDS];
+const long blinkInterval = 300; 
+const long fireUpdate = 20;
+unsigned long previousMillis = 0; 
+CRGBPalette16 gPal;
+
+void setupFASTLED() {
+  trc(F("FASTLED_DATA_PIN "));
+  trc(String(FASTLED_DATA_PIN));
+  trc(F("FASTLED_NUM_LEDS "));
+  trc(String(FASTLED_NUM_LEDS));
+  trc(F("ZactuatorFASTLED setup done "));
+  FastLED.addLeds<FASTLED_TYPE, FASTLED_DATA_PIN>(leds, FASTLED_NUM_LEDS);
+}
+
+
+
+void FASTLEDLoop()
+{
+   if (currentLEDState == GENERAL)
+   {
+    unsigned long currentMillis = millis();
+    if (currentMillis - previousMillis >= blinkInterval) 
+    {
+      // save the last time you blinked the LED
+      previousMillis = currentMillis;
```

Comment:
this will not give a constant blink interval

Reply:
I have no better idea at the moment. Do you have a better idea?

Reply:
```C++
long lastUpdate = 0;
long currentUpdate = 0;

//returns the current step of the animation
int animation_step(int duration, int steps) {
	int currentStep = ((currentUpdate % duration) / ((float)duration)) * steps;
	return currentStep;
}

//returns the number of steps since the last update
int animation_step_count(int duration, int steps) {
	long lastAnimationNumber = lastUpdate / duration;
	long currentAnimationNumber = currentUpdate / duration;
	int lastStep = ((lastUpdate % duration) / ((float)duration)) * steps;
	int currentStep = ((currentUpdate % duration) / ((float)duration)) * steps;

	return currentStep - lastStep + (currentAnimationNumber - lastAnimationNumber) * steps;
}

void loop() {
	lastUpdate = currentUpdate;
	currentUpdate = millis();

	// blink led
	int count = animation_step_count(600, 2);
	if (count > 0) {
		int step = animation_step(duration, 2);
		if (step == 0) {
			//set led on
		} else {
			//set led off
		}
	}
}
```

the `animation_step` and `animation_step_count` are from one of my other projects

Reply:
you can get the current animation step, animation steps since you last updated the leds and you can also detect if the animation finished since last updated the leds. You only have to specify the duration and the number of steps you animation should have. The animations run in a loop, so after `duration` milliseconds the animation step start by 0 again.

Reply:
should be "600" of `animation_step_count(600,2)`; the same like _duration_?

Reply:
yes the first argument is the duration and the second the step count

Reply:
ah yes i see, updated my comment

Reply:
I updated "my" code...